File: Algo/Algorithm/Questions/Arrays/22.waterfallStreams/main.js
// O(w^2 * h) time | O(w) space - where w and h are the width and height of the input array.
function waterfallStreams(array, source) {

  let prevRow = [...array[0]]
  prevRow[source] = -1

  for (let i = 1; i < array.length; i++) {
    const currRow = [...array[i]]
    for (let j = 0; j < currRow.length; j++) {

      const valueAbove = prevRow[j]
      const hasWatherAbove = valueAbove < 0
      const hasBlock = currRow[j] === 1

      if (!hasWatherAbove) {
        continue
      }

      if (!hasBlock) {
        currRow[j] += valueAbove
        continue
      }

      const splitWater = valueAbove / 2

      let leftIndx = j
      while (leftIndx - 1 >= 0) {
        leftIndx--
        if (prevRow[leftIndx] === 1) {
          break
        }

        if (currRow[leftIndx] !== 1) {
          currRow[leftIndx] += splitWater
          break
        }
      }

      let rightIndx = j
      while (rightIndx + 1 < currRow.length) {
        rightIndx++
        if (prevRow[rightIndx] === 1) {
          break
        }

        if (currRow[rightIndx] !== 1) {
          currRow[rightIndx] += splitWater
          break
        }
      }
    }
    prevRow = currRow
  }
  return prevRow.map(v => v < 0 ? v * -100 : v)
}

File: Algo/Algorithm/Questions/Arrays/22.waterfallStreams/main.cpp
#include <vector>
#include <iostream>

using namespace std;

vector<double> waterfallStreams(vector<vector<double>> array, int source)
{
  int len = array[0].size();
  vector<double> rowAbove(len);
  rowAbove[source] = 100;

  for (int row = 0; row < array.size(); row++)
  {
    vector<double> currRow(len);
    for (int col = 0; col < len; col++)
    {
      double hasWatherAbove = rowAbove[col];
      bool noBlocker = array[row][col] == 0;
      if (hasWatherAbove)
      {
        if (noBlocker)
          currRow[col] += rowAbove[col];
        else
        {
          int rightIndx = col;
          while (rightIndx < len && array[row - 1][rightIndx] != 1)
          {
            if (array[row][rightIndx] == 0)
            {
              currRow[rightIndx] += (rowAbove[col] / 2);
              break;
            }
            rightIndx++;
          }

          int leftIndx = col;
          while (leftIndx >= 0 && array[row - 1][leftIndx] != 1)
          {
            if (array[row][leftIndx] == 0)
            {
              currRow[leftIndx] += (rowAbove[col] / 2);
              break;
            }
            leftIndx--;
          }
        }
      }
    }
    rowAbove = currRow;
  }
  // Write your code here.
  return rowAbove;
}

int main()
{
  vector<vector<double>> array = {
      {0, 0, 0, 0, 0, 0, 0},
      {1, 0, 0, 0, 0, 0, 0},
      {0, 0, 1, 1, 1, 0, 0},
      {0, 0, 0, 0, 0, 0, 0},
      {1, 1, 1, 0, 0, 1, 0},
      {0, 0, 0, 0, 0, 0, 1},
      {0, 0, 0, 0, 0, 0, 0}};
  int source = 3;
  for (double x : waterfallStreams(array, source))
    cout << x << " ";
  cout << endl;
  return 0;
}